fix compilation for musl

a16bitsysop opened the pull request:
musl lacks execinfo.h
lfs64 definitions are now deprecated and will be removed in the next release of musl

Fixes #1743

a16bitsysop commented:
not sure if the config.hh creation is correct and also if `__USE_FILE_OFFSET64` is correct

a16bitsysop commented:
removed adding `ucontext` to `target_link_libraries` for `seastar` as fails to link on alpine/musl without it, but fails on glibc with it.  Guess it needs a check to decide if needed?

tchaikov reviewed (commented):
@a16bitsysop hi Duncan, much appreciated!

i'd suggest split it into smaller commits, each of which is accompanied with its own explanation. this would be non-trivial.  i can help on this. but i am not able to allocate large chunk of time on this. so if you are fine with this, we can divide and conquer this problem -- i can extract part of this into a separated PR. and leave the rest to you. and if you don't have enough bandwidth, i can go on with the rest. but if you do, that'd be great!

tchaikov commented on CMakeLists.txt:
```diff
@@ -408,6 +408,10 @@ include (CheckIncludeFileCXX)
+pkg_check_modules (UCONTEXT IMPORTED_TARGET libucontext)
+if (UCONTEXT_FOUND)
+  set (UCONTEXT_LIB PkgConfig::UCONTEXT)
+endif ()
```
i'd extract this into a separate `Finducontext.cmake`. it would resemble `Findrt.cmake`.

a16bitsysop replied:
okay will have a look

a16bitsysop replied:
I will push with the changes in new PR removed (I rebased it onto this branch)

tchaikov replied:
also, the introduction of `config.hh` is an important design decision. currently, we are using macros to persist various settings, and to pass them down to the Seastar applications. i think both of them have pros and cons. personally, i am in favor of `config.hh`. but i think we need more inputs on this regard. as this would be a tree-wide change if we want to be consistent on persisting / populating the configurations.

a16bitsysop replied:
okay, I wasn’t sure how to pass the defines down into the source code from cmake

a16bitsysop commented:
> @a16bitsysop hi Duncan, much appreciated!
> 
> i'd suggest split it into smaller commits, each of which is accompanied with its own explanation. this would be non-trivial. i can help on this. but i am not able to allocate large chunk of time on this. so if you are fine with this, we can divide and conquer this problem -- i can extract part of this into a separated PR. and leave the rest to you. and if you don't have enough bandwidth, i can go on with the rest. but if you do, that'd be great!

Okay, will split them up.  I could create a separate PR from the original patches by @neopskose with the stderr stuff as that is alot of changes

tchaikov commented on src/core/file.cc:
```diff
@@ -383,8 +383,14 @@ posix_file_impl::list_directory(std::function<future<> (directory_entry de)> nex
+	off64_t        d_off; /* 64-bit offset to next structure */
+#else
+        ino_t          d_ino; /* 64-bit inode number */
+        off_t          d_off; /* 64-bit offset to next structure */
```
i think musl libc has not dropped the legacy types for large file support. see https://git.musl-libc.org/cgit/musl/tree/include/dirent.h#n59 . they are not provided by `_GNU_SOURCE` feature macro anymore, but under `_LARGEFILE64_SOURCE`. but yeah, they will be eventually removed. see https://musl.libc.org/releases.html for the release note for v1.2.4.

so, in the light of that `ino64_t` and `off64_t` are glibc extensions, and they are defined when `__USE_LARGEFILE64` is defined. and glibc also has type aliases of `ino_t` and `off_t` for these types if `_FILE_OFFSET_BITS` is 64. this macro is set by user.

`__USE_FILE_OFFSET64` is used by the glibc internally, it is defined when `_FILE_OFFSET_BITS` is 64.  if `__USE_FILE_OFFSET64` is defined, glibc defines 64bit APIs as default. which means the types like `ino_t` would be 64-bit, and `readdir()` would be 64-bit as well. one don't have to use the `readdir64()` variant explicitly.

so,

1. we should avoid using a macro used by glibc internally as a part of its public interface.
2. and the guarded statements are not quite relevant. as we should be able to use `ino_t` in the "64-bit" mode. as we want to use 64-bit inode number and offset.

as IMHO, the 64-bit inode number and 64-bit offset should be the default in this context as Seastar is designed to be used to build high-performance services, and they should be default under most circumstances. the existence their 32-bit cousins is only for the 32-bit systems -- only for embedded and some customized systems.

so, i think it'd be better to use `ino_t` and `off_t` directly. to achieve this goal, we can just put
```c++
static_assert(sizeof(ino_t) == 8, "large file support not enabled");
static_assert(sizeof(off_t) == 8, "large file support not enabled");
```

as by default, if we compile the tree with gcc or clang, `_GNU_SOURCE` is always defined.  and libstdc++ actually needs this, see https://gcc.gnu.org/onlinedocs/libstdc++/faq.html#faq.predefined. and glibc would define `_LARGEFILE64_SOURCE` if this macro is defined. and musl libc just use the 64-bit `ino_t` and `off_t`. so we should be safe in both cases without defining `_FILE_OFFSET_BITS=64` explicitly.

a16bitsysop replied:
okay will chenage

a16bitsysop commented:
updated and dropped `sys/sdt.h` check for the minute as that needs to check for systemtap, changed passing execinfo status to target_compile_definitions

ucontext check is copied from rt check, but it doesn’t work/run and I can’t see why

tchaikov commented on CMakeLists.txt:
```diff
@@ -796,6 +800,7 @@ target_link_libraries (seastar
     StdAtomic::atomic
     lksctp-tools::lksctp-tools
     rt::rt
+    ucontext:ucontext
```
@a16bitsysop you might need to put `ucontext::ucontext` (two colons). but i created #1799, as i think there is no need to use a dedicated source file for checking the availability of the `getcontext()` function. could you please give it a try ?

a16bitsysop replied:
thanks, dropped that commit and tested yours

tchaikov commented:
> updated and dropped `sys/sdt.h` check for the minute as that needs to check for systemtap, changed passing execinfo status to target_compile_definitions
> 

i created #1823 to address this. could you help review / test it ?

tchaikov commented on include/seastar/util/backtrace.hh:
```diff
@@ -59,13 +61,16 @@ frame decorate(uintptr_t addr) noexcept;
 SEASTAR_MODULE_EXPORT
 template<typename Func>
 void backtrace(Func&& func) noexcept(noexcept(func(frame()))) {
+    #ifdef SEASTAR_HAVE_EXECINFO
```
we don't indent `#ifdef` in this project elsewhere. so probably we don't need to do so here?

a16bitsysop replied:
okay changed

tchaikov commented on include/seastar/util/backtrace.hh:
```diff
@@ -59,13 +61,16 @@ frame decorate(uintptr_t addr) noexcept;
         func(decorate(ip - 1));
     }
+    #endif
+    return;
```
this `return` is not necessary.

a16bitsysop replied:
yes works without

tchaikov commented on src/util/backtrace.cc:
```diff
@@ -22,7 +22,9 @@
 
+#ifdef SEASTAR_HAVE_EXECINFO
 #include <execinfo.h>
+#endif
```
i think we can drop this `#include`. as this source file does not reference the symbols exposed by `execinfo.h`

a16bitsysop replied:
dropped it

tchaikov commented on src/util/backtrace.cc:
```diff
@@ -95,11 +97,13 @@ frame decorate(uintptr_t addr) noexcept {
 
 simple_backtrace current_backtrace_tasklocal() noexcept {
     simple_backtrace::vector_type v;
+    #ifdef SEASTAR_HAVE_EXECINFO
```
no need to `#ifdef` here, i think. as `backtrace()` would be a no-op if execinfo is not available.

a16bitsysop replied:
changed

tchaikov commented on CMakeLists.txt:
```diff
@@ -870,6 +874,11 @@ if (Seastar_SSTRING)
     PUBLIC SEASTAR_SSTRING)
 endif ()
 
+if (SEASTAR_HAVE_EXECINFO)
```
i'd move the `include()` and `check_include_file_cxx ()` call closer to where `SEASTAR_HAVE_EXECINFO` is used.

tchaikov replied:
also, we tend to use `Seastar_HAVE_EXECINFO`  as variable name in CMake.

a16bitsysop replied:
okay

tchaikov replied:
> i'd move the `include()` and `check_include_file_cxx ()` call closer to where `SEASTAR_HAVE_EXECINFO` is used.

@a16bitsysop could you reconsider this suggestion?

a16bitsysop replied:
I moved after seastar target is declared, it can’t go before as otherwise target doesn’t exist

tchaikov commented on CMakeLists.txt:
```diff
@@ -870,6 +874,11 @@ if (Seastar_SSTRING)
 
+if (SEASTAR_HAVE_EXECINFO)
+target_compile_definitions (seastar                                      
+	PUBLIC SEASTAR_HAVE_EXECINFO)
```
could you use two spaces for indent here?

a16bitsysop replied:
changed

tchaikov commented on CMakeLists.txt:
```diff
@@ -405,6 +405,10 @@ find_package (Threads REQUIRED)
 
+include (CheckIncludeFileCXX)
+check_include_file_cxx (execinfo.h SEASTAR_HAVE_EXECINFO)
+#configure_file (config.hh.in ${CMAKE_CURRENT_SOURCE_DIR}/include/seastar/config.hh @ONLY)
```
might want to drop commented out code.

tchaikov commented on CMakeLists.txt:
```diff
@@ -750,6 +753,11 @@ add_library (seastar
 
 add_library (Seastar::seastar ALIAS seastar)
 
+if (Seastar_HAVE_EXECINFO)
+target_compile_definitions (seastar
+  PUBLIC Seastar_HAVE_EXECINFO)
+endif()
```
```suggestion
if (Seastar_HAVE_EXECINFO)
  target_compile_definitions (seastar
    PUBLIC Seastar_HAVE_EXECINFO)
endif()
```

tchaikov commented on CMakeLists.txt:
```diff
@@ -754,9 +754,9 @@ add_library (seastar
-endif()
+  target_compile_definitions (seastar
+    PUBLIC Seastar_HAVE_EXECINFO)
+endif ()
```
could you please squash this change into the one which introduced `Seastar_HAVE_EXECINFO`?

a16bitsysop replied:
okay, squashed

tchaikov commented on src/core/file.cc:
```diff
@@ -382,9 +382,13 @@ posix_file_impl::list_directory(std::function<future<> (directory_entry de)> nex
     // instead.
 
     // From getdents(2):
+    /* check for 64-bit inode number */
```
nit, might want to put
```c++
// check for 64-bit inode number
```

tchaikov reviewed (approved):
lgtm.

tchaikov commented on tests/unit/unwind_test.cc:
```diff
@@ -50,12 +50,14 @@ BOOST_AUTO_TEST_CASE(test_signal_mask_is_preserved_on_unwinding) {
     }
 
-    // Check backtrace()
+    // Check backtrace() unless execinfo.h is not present
```
```suggestion
    // Check backtrace() unless execinfo.h is present
```

tchaikov commented on tests/unit/unwind_test.cc:
```diff
@@ -50,7 +50,7 @@ BOOST_AUTO_TEST_CASE(test_signal_mask_is_preserved_on_unwinding) {
     }
 
-    // Check backtrace() unless execinfo.h is not present
+    // Check backtrace() if execinfo.h is present
```
could you please squash this change as well?

a16bitsysop replied:
done it

tchaikov reviewed (approved):
lgtm

tchaikov commented:
@scylladb/seastar-maint hello maintainers, could you help review this change?

nyh commented on include/seastar/util/backtrace.hh:
```diff
@@ -28,7 +28,9 @@
 #include <seastar/util/modules.hh>
 
 #ifndef SEASTAR_MODULE
+#ifdef Seastar_HAVE_EXECINFO
```
This is so much cmake boilerplate code just to check if <execinfo.h> exists in nothing more. Since C++17, it's no longer needed - "__has_include" exists. Can we just use that and avoid all cmake changes?
since you need to know also below if execinfo exists, you can just do something like
```
#if __has_include(<execinfo.h>)
#include <execinfo.h>
#define HAVE_EXECINFO
#endif
```
and below check HAVE_EXECINFO.

nyh commented on include/seastar/util/backtrace.hh:
```diff
@@ -59,13 +61,15 @@ frame decorate(uintptr_t addr) noexcept;
         auto ip = reinterpret_cast<uintptr_t>(buffer[i]);
         func(decorate(ip - 1));
     }
+#endif
```
I wonder if we shouldn't in an `#else` return some sort of string saying backtrace() is unimplemented. Or at least leave a "FIXME" comment. One day we'll need to fix that, right?

nyh replied:
By the way, a question about the preexisting code: Why is backtrace() defined in the header file? Do end users use it directly, or things implemented with it? Even if end-users do use it directly, it should switch to taking std::function (or seastar::noncopyable_function) instead of being a template - and then the implementation can be moded to the source file. It's not important for performance.

nyh commented on tests/unit/exception_logging_test.cc:
```diff
@@ -200,9 +200,11 @@ BOOST_AUTO_TEST_CASE(throw_with_backtrace_exception_logging) {
         log_msg << std::current_exception();
     }
 
+#ifdef Seastar_HAVE_EXECINFO
```
In the long run, this ifdef is wrong: Eventually we'll want to implement a backtrace correctly even if HAVE_EXECINFO is false, using some other method.
One way you can solve this is to do in backtrace.hh something like:
```c++
#if __has_include(<execinfo.h>)
#include <execinfo.h>
#define HAVE_EXECINFO
#endif
...
backtrace() {
#ifdef HAVE_EXECINFO
... some real definition
#else
   // Not implemented yet
#define SEASTAR_BACKTRACE_UNIMPLEMENTED
#endif
```

Then in the test you can check for `#ifndef SEASTAR_BACKTRACE_UNIMPLEMENTED`. In the future, we can implement it, and remove this macro, even when execinfo.h doesn't exist.

a16bitsysop replied:
Yes that is simpler

nyh reviewed (commented):
Looks good, but I had some ideas on how to make the backtrace stuff simpler and more forward-looking to when we have backtrace() impelemented in any case. For your consideration.
